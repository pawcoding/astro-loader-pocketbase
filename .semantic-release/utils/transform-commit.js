/**
 * Transform function for semantic-release commit processing
 * Categorizes commits and processes issue/PR links
 */
function transformCommit(commit, context) {
  const issues = [];

  commit.notes.forEach((note) => {
    note.title = "BREAKING CHANGES";
  });

  // Group different commit types under appropriate categories
  if (commit.type === "feat") {
    commit.type = "Features";
  } else if (commit.type === "fix") {
    commit.type = "Bug Fixes";
  } else if (commit.type === "docs") {
    commit.type = "Documentation";
  } else if (
    ["style", "refactor", "test", "build", "ci", "chore"].includes(commit.type)
  ) {
    commit.type = "Internals";
  } else if (commit.type === "perf") {
    commit.type = "Performance Improvements";
  } else if (commit.revert) {
    commit.type = "Reverts";
  } else {
    return;
  }

  if (commit.scope === "*") {
    commit.scope = "";
  }

  if (typeof commit.hash === "string") {
    commit.shortHash = commit.hash.substring(0, 7);
  }

  if (typeof commit.subject === "string") {
    let url = context.repository
      ? `${context.host}/${context.owner}/${context.repository}`
      : context.repoUrl;
    if (url) {
      url = `${url}/issues/`;
      // Issue URLs.
      commit.subject = commit.subject.replace(/#([0-9]+)/g, (_, issue) => {
        issues.push(issue);
        return `[#${issue}](${url}${issue})`;
      });
    }
    if (context.host) {
      // User URLs.
      commit.subject = commit.subject.replace(
        /\B@([a-z0-9](?:-?[a-z0-9/]){0,38})/g,
        (_, username) => {
          if (username.includes("/")) {
            return `@${username}`;
          }

          return `[@${username}](${context.host}/${username})`;
        }
      );
    }
  }

  // remove references that already appear in the subject
  commit.references = commit.references.filter((reference) => {
    if (issues.indexOf(reference.issue) === -1) {
      return true;
    }

    return false;
  });

  return commit;
}

module.exports = { transformCommit };
