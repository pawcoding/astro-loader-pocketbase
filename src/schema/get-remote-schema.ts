import { pocketbaseErrorResponseSchema } from "../types/pocketbase-api-responses.type";
import type { PocketBaseLoaderOptions } from "../types/pocketbase-loader-options.type";
import type { PocketBaseCollection } from "../types/pocketbase-schema.type";
import { pocketBaseCollection } from "../types/pocketbase-schema.type";

/**
 * Fetches the schema for the specified collection from the PocketBase instance.
 *
 * @param options Options for the loader. See {@link PocketBaseLoaderOptions} for more details.
 * @param token The superuser token to authenticate the request.
 */
export async function getRemoteSchema(
  options: Pick<PocketBaseLoaderOptions, "collectionName" | "url">,
  token: string
): Promise<PocketBaseCollection | undefined> {
  // Build URL and headers for the schema request
  const schemaUrl = new URL(
    `api/collections/${options.collectionName}`,
    options.url
  ).href;
  const schemaHeaders = new Headers();
  schemaHeaders.set("Authorization", token);

  // Fetch the schema
  const schemaRequest = await fetch(schemaUrl, {
    headers: schemaHeaders
  });

  // If the request was not successful, try another method
  if (!schemaRequest.ok) {
    try {
      const errorData = (await schemaRequest.json()) as unknown;
      const parsedError = pocketbaseErrorResponseSchema.parse(errorData);
      const errorMessage = `Fetching schema from ${options.collectionName} failed with status code ${schemaRequest.status}.\nReason: ${parsedError.message}`;
      console.error(errorMessage);
    } catch {
      // If parsing fails, just log the status code
      console.error(
        `Fetching schema from ${options.collectionName} failed with status code ${schemaRequest.status}.`
      );
    }

    return undefined;
  }

  // Get the schema from the response
  const schemaData = (await schemaRequest.json()) as unknown;
  // Parse and validate the schema using the PocketBase collection schema
  return pocketBaseCollection.parse(schemaData);
}
