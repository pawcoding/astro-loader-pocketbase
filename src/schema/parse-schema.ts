import { z } from "astro/zod";
import type {
  PocketBaseCollection,
  PocketBaseSchemaEntry
} from "../types/pocketbase-schema.type";

/**
 * Converts PocketBase collection fields into Zod types, handling field types, required status, and custom schemas.
 */
export function parseSchema(
  collection: PocketBaseCollection,
  customSchemas: Record<string, z.ZodType> | undefined,
  hasSuperuserRights: boolean,
  improveTypes: boolean,
  experimentalLiveTypesOnly?: boolean,
  fieldsToInclude?: Array<string>
): Record<string, z.ZodType> {
  // Prepare the schemas fields
  const fields: Record<string, z.ZodType> = {};

  // Parse every field in the schema
  for (const field of collection.fields) {
    // Skip hidden fields if the user does not have superuser rights
    if (field.hidden && !hasSuperuserRights) {
      continue;
    }

    // If fieldsToInclude is specified, only include fields that are in the list
    if (fieldsToInclude && !fieldsToInclude.includes(field.name)) {
      continue;
    }

    let fieldType;

    // Determine the field type and create the corresponding Zod type
    switch (field.type) {
      case "number":
        fieldType = z.number();
        break;
      case "bool":
        fieldType = z.boolean();
        break;
      case "date":
      case "autodate":
        if (experimentalLiveTypesOnly) {
          // If experimental live types only mode is enabled, treat dates as strings
          fieldType = z.string();
          break;
        }
        // Coerce and parse the value as a date
        fieldType = z.coerce.date();
        break;
      case "geoPoint":
        fieldType = z.object({
          lon: z.number(),
          lat: z.number()
        });
        break;
      case "select": {
        if (!field.values) {
          throw new Error(
            `Field ${field.name} is of type "select" but has no values defined.`
          );
        }

        // Create an enum for the select values
        // @ts-expect-error - Zod complains because the values are not known at compile time and thus the array is not static.
        const values = z.enum(field.values);

        // Parse the field type based on the number of values it can have
        fieldType = parseSingleOrMultipleValues(field, values);
        break;
      }
      case "relation":
      case "file":
        // NOTE: Relations are currently not supported and are treated as strings
        // NOTE: Files are later transformed to URLs

        // Parse the field type based on the number of values it can have
        fieldType = parseSingleOrMultipleValues(field, z.string());
        break;
      case "json":
        if (customSchemas && customSchemas[field.name]) {
          // Use the user defined custom schema for the field
          fieldType = customSchemas[field.name];
        } else {
          // Parse the field as unknown JSON
          fieldType = z.unknown();
        }
        break;
      default:
        // Default to a string
        fieldType = z.string();
        break;
    }

    const isRequired =
      // Check if the field is required
      field.required ||
      // `onCreate autodate` fields are always set
      (field.type === "autodate" && field.onCreate) ||
      // Improve number and boolean types by providing default values
      (improveTypes && (field.type === "number" || field.type === "bool"));

    // If the field is not required, mark it as optional
    if (!isRequired) {
      fieldType = z.preprocess(
        (val) => val || undefined,
        z.optional(fieldType)
      );
    }

    // Add the field to the fields object
    fields[field.name] = fieldType;
  }

  return fields;
}

/**
 * Parse the field type based on the number of values it can have
 *
 * @param field Field to parse
 * @param type Type of each value
 *
 * @returns The parsed field type
 */
function parseSingleOrMultipleValues(
  field: PocketBaseSchemaEntry,
  type: z.ZodType
): z.ZodType {
  // If the select allows multiple values, create an array of the enum
  if (field.maxSelect === undefined || field.maxSelect === 1) {
    return type;
  }

  return z.array(type);
}
