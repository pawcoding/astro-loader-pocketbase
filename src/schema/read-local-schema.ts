import fs from "fs/promises";
import path from "path";
import type { PocketBaseCollection } from "../types/pocketbase-schema.type";

/**
 * Reads the local PocketBase schema file and returns the schema for the specified collection.
 *
 * @param localSchemaPath Path to the local schema file.
 * @param collectionName Name of the collection to get the schema for.
 */
export async function readLocalSchema(
  localSchemaPath: string,
  collectionName: string
): Promise<PocketBaseCollection | undefined> {
  const realPath = path.join(process.cwd(), localSchemaPath);

  try {
    // Read the schema file
    const schemaFile = await fs.readFile(realPath, "utf-8");
    const database: Array<PocketBaseCollection> = JSON.parse(schemaFile);

    // Check if the database file is valid
    if (!database || !Array.isArray(database)) {
      throw new Error("Invalid schema file");
    }

    // Find and return the schema for the collection
    const schema = database.find(
      (collection) => collection.name === collectionName
    );

    if (!schema) {
      throw new Error(
        `Collection "${collectionName}" not found in schema file`
      );
    }

    return schema;
  } catch (error) {
    console.error(
      `Failed to read local schema from ${localSchemaPath}. No types will be generated.\nReason: ${error}`
    );
    return undefined;
  }
}
