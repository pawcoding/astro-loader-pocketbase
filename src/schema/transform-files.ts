import type { PocketBaseEntry } from "../types/pocketbase-entry.type";
import type { PocketBaseSchemaEntry } from "../types/pocketbase-schema.type";

/**
 * Transforms file names in a PocketBase entry to file URLs.
 *
 * @param baseUrl URL of the PocketBase instance.
 * @param collection Collection of the entry.
 * @param entry Entry to transform.
 */
export function transformFiles(
  baseUrl: string,
  fileFields: Array<PocketBaseSchemaEntry>,
  entry: PocketBaseEntry
): PocketBaseEntry {
  // Transform all file names to file URLs
  for (const field of fileFields) {
    const fieldName = field.name;

    if (field.maxSelect === 1) {
      const fileName = entry[fieldName] as string | undefined;
      // Check if a file name is present
      if (!fileName) {
        continue;
      }

      // Transform the file name to a file URL
      entry[fieldName] = transformFileUrl(
        baseUrl,
        entry.collectionName,
        entry.id,
        fileName
      );
    } else {
      const fileNames = entry[fieldName] as Array<string> | undefined;
      // Check if file names are present
      if (!fileNames) {
        continue;
      }

      // Transform all file names to file URLs
      entry[fieldName] = fileNames.map((file) =>
        transformFileUrl(baseUrl, entry.collectionName, entry.id, file)
      );
    }
  }

  return entry;
}

/**
 * Transforms a file name to a PocketBase file URL.
 *
 * @param base Base URL of the PocketBase instance.
 * @param collectionName Name of the collection.
 * @param entryId ID of the entry.
 * @param file Name of the file.
 */
function transformFileUrl(
  base: string,
  collectionName: string,
  entryId: string,
  file: string
): string {
  return `${base}/api/files/${collectionName}/${entryId}/${file}`;
}
