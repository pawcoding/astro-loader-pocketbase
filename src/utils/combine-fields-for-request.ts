import type {
  PocketBaseLoaderBaseOptions,
  PocketBaseLoaderOptions
} from "../types/pocketbase-loader-options.type";

/**
 * Combine basic, special, and user-specified fields for PocketBase API requests.
 * This utility ensures that required system fields are always included in API requests.
 *
 * @param userFields Array of fields specified by the user, or undefined for all fields
 * @param options PocketBase loader options containing custom field configurations
 * @returns Combined array of fields to include in the API request, or undefined for all fields
 */
export function combineFieldsForRequest(
  userFields: Array<string> | undefined,
  options: Pick<PocketBaseLoaderBaseOptions, "updatedField" | "contentFields"> &
    Pick<PocketBaseLoaderOptions, "idField">
): Array<string> | undefined {
  // If no fields specified, return undefined to get all fields
  if (!userFields) {
    return undefined;
  }

  // Basic fields that are always required by the loader
  const basicFields = ["id", "collectionId", "collectionName"];

  // Special fields that are configured in options
  const specialFields: Array<string> = [];

  // Add custom id field if specified
  if (options.idField && options.idField !== "id") {
    specialFields.push(options.idField);
  }

  // Add updated field if specified
  if (options.updatedField) {
    specialFields.push(options.updatedField);
  }

  // Add content fields if specified
  if (options.contentFields) {
    if (Array.isArray(options.contentFields)) {
      specialFields.push(...options.contentFields);
    } else {
      specialFields.push(options.contentFields);
    }
  }

  // Combine all field sets, removing duplicates
  const allFields = [
    ...new Set([...basicFields, ...specialFields, ...userFields])
  ];

  return allFields;
}
