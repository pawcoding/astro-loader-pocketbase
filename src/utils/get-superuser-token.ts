import type { AstroIntegrationLogger } from "astro";

/**
 * This function will get a superuser token from the given PocketBase instance.
 *
 * @param url URL of the PocketBase instance
 * @param superuserCredentials Credentials of the superuser
 *
 * @returns A superuser token to access all resources of the PocketBase instance.
 */
export async function getSuperuserToken(
  url: string,
  superuserCredentials: {
    email: string;
    password: string;
  },
  logger?: AstroIntegrationLogger
): Promise<string | undefined> {
  // Build the URL for the login endpoint
  const loginUrl = new URL(
    `api/collections/_superusers/auth-with-password`,
    url
  ).href;

  // Create a new FormData object to send the login data
  const loginData = new FormData();
  loginData.set("identity", superuserCredentials.email);
  loginData.set("password", superuserCredentials.password);

  // Send the login request to get a token
  const loginRequest = await fetch(loginUrl, {
    method: "POST",
    body: loginData
  });

  // If the login request was not successful, print the error message and return undefined
  if (!loginRequest.ok) {
    if (loginRequest.status === 429) {
      const info =
        "A rate limit was hit while trying to authenticate with PocketBase. Consider using an `impersonateToken` as credentials to avoid this issue.";
      if (logger) {
        logger.info(info);
      } else {
        console.info(info);
      }

      // Random wait between 3 (default rate limit interval) and 8 seconds
      const retryAfter = Math.random() * 5 + 3;
      // oxlint-disable-next-line promise/avoid-new
      await new Promise((resolve) => setTimeout(resolve, retryAfter * 1000));

      return getSuperuserToken(url, superuserCredentials, logger);
    }

    const reason = await loginRequest.json().then((data) => data.message);
    const errorMessage = `The given email / password for ${url} was not correct. Astro can't generate type definitions automatically and may not have access to all resources.\nReason: ${reason}`;
    if (logger) {
      logger.error(errorMessage);
    } else {
      console.error(errorMessage);
    }
    return undefined;
  }

  // Return the token
  return await loginRequest.json().then((data) => data.token);
}
