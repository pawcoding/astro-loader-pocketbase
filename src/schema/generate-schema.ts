import type { ZodObject, ZodSchema } from "astro/zod";
import { z } from "astro/zod";
import type {
  ExpandedFields,
  PocketBaseCollection
} from "../types/pocketbase-collection.type";
import type { PocketBaseLoaderOptions } from "../types/pocketbase-loader-options.type";
import { getRemoteSchema } from "./get-remote-schema";
import { parseSchema } from "./parse-schema";
import { readLocalSchema } from "./read-local-schema";
import { transformFiles } from "./transform-files";

/**
 * Basic schema for every PocketBase collection.
 */
const BASIC_SCHEMA = {
  id: z.string(),
  collectionId: z.string(),
  collectionName: z.string()
};

/**
 * Types of fields that can be used as an ID.
 */
const VALID_ID_TYPES = ["text", "number", "email", "url", "date"];

/**
 * Generate a schema for the collection based on the collection's schema in PocketBase.
 * By default, a basic schema is returned if no other schema is available.
 * If superuser credentials are provided, the schema is fetched from the PocketBase API.
 * If a path to a local schema file is provided, the schema is read from the file.
 *
 * @param options Options for the loader. See {@link PocketBaseLoaderOptions} for more details.
 */
export async function generateSchema(
  options: PocketBaseLoaderOptions
): Promise<ZodSchema> {
  let collection: PocketBaseCollection | undefined;
  const expandedFields: ExpandedFields = {};

  // Try to get the schema directly from the PocketBase instance
  collection = await getRemoteSchema(options);

  const hasSuperuserRights = !!collection || !!options.superuserCredentials;

  // If the schema is not available, try to read it from a local schema file
  if (!collection && options.localSchema) {
    collection = await readLocalSchema(
      options.localSchema,
      options.collectionName
    );
  }

  // If the schema is still not available, return the basic schema
  if (!collection) {
    console.error(
      `No schema available for "${options.collectionName}". Only basic types are available. Please check your configuration and provide a valid schema file or superuser credentials.`
    );
    // Return the basic schema since every collection has at least these fields
    return z.object(BASIC_SCHEMA);
  }

  // Parse the schema
  const fields = parseSchema(
    collection,
    options.jsonSchemas,
    hasSuperuserRights,
    options.improveTypes ?? false
  );

  // Check if custom id field is present
  if (options.idField) {
    // Find the id field in the schema
    const idField = collection.fields.find(
      (field) => field.name === options.idField
    );

    // Check if the id field is present and of a valid type
    if (!idField) {
      console.error(
        `The id field "${options.idField}" is not present in the schema of the collection "${options.collectionName}".`
      );
    } else if (!VALID_ID_TYPES.includes(idField.type)) {
      console.error(
        `The id field "${options.idField}" for collection "${
          options.collectionName
        }" is of type "${
          idField.type
        }" which is not recommended. Please use one of the following types: ${VALID_ID_TYPES.join(
          ", "
        )}.`
      );
    }
  }

  // Check if the content field is present
  if (
    typeof options.contentFields === "string" &&
    !fields[options.contentFields]
  ) {
    console.error(
      `The content field "${options.contentFields}" is not present in the schema of the collection "${options.collectionName}".`
    );
  } else if (Array.isArray(options.contentFields)) {
    for (const field of options.contentFields) {
      if (!fields[field]) {
        console.error(
          `The content field "${field}" is not present in the schema of the collection "${options.collectionName}".`
        );
      }
    }
  }

  // Check if the updated field is present
  if (options.updatedField) {
    if (!fields[options.updatedField]) {
      console.error(
        `The field "${options.updatedField}" is not present in the schema of the collection "${options.collectionName}".\nThis will lead to errors when trying to fetch only updated entries.`
      );
    } else {
      const updatedField = collection.fields.find(
        (field) => field.name === options.updatedField
      );
      if (
        !updatedField ||
        updatedField.type !== "autodate" ||
        !updatedField.onUpdate
      ) {
        console.warn(
          `The field "${options.updatedField}" is not of type "autodate" with the value "Update" or "Create/Update".\nMake sure that the field is automatically updated when the entry is updated!`
        );
      }
    }
  }

  if (options.expand) {
    for (const expandedFieldName of options.expand) {
      const expandedFieldDefinition = collection.fields.find(
        (field) => field.name === expandedFieldName
      );

      if (!expandedFieldDefinition) {
        console.error(
          `The provided field in the expand property "${expandedFieldName}" is not present in the schema of the collection "${options.collectionName}".\nThis will lead to use unable to provide a definition for this field.`
        );
      } else {
        if (!expandedFieldDefinition.collectionId) {
          console.error(
            `The provided field in the expand property "${expandedFieldName}" does not have an associated collection linked to it, we need this in order to know the shape of the related schema.`
          );
        } else {
          const expandedSchema = await generateSchema({
            collectionName: expandedFieldDefinition.collectionId,
            superuserCredentials: options.superuserCredentials,
            localSchema: options.localSchema,
            jsonSchemas: options.jsonSchemas,
            improveTypes: options.improveTypes,
            url: options.url
          });

          expandedFields[expandedFieldName] = z.union([
            expandedchema,
            z.array(expandedchema)
          ]);
        }
      }
    }
  }

  const expandSchema = {
    expand: buildExpandSchema(expandedFields).optional()
  };

  const schema = z.object({
    ...BASIC_SCHEMA,
    ...fields,
    ...expandSchema
  });

  // Get all file fields
  const fileFields = collection.fields
    .filter((field) => field.type === "file")
    // Only show hidden fields if the user has superuser rights
    .filter((field) => !field.hidden || hasSuperuserRights);

  if (fileFields.length === 0) {
    return schema;
  }

  // Transform file names to file urls
  return schema.transform((entry) => {
    if (Array.isArray(entry)) {
      return entry.map((e) => transformFiles(options.url, fileFields, e));
    }

    return transformFiles(options.url, fileFields, entry);
  });
}

/**
 * Builds a Zod object schema from expandedFields, where each property is either a ZodSchema or an array of ZodSchema.
 */
export function buildExpandSchema(
  expandedFields: ExpandedFields
): ZodObject<Record<string, ZodSchema | z.ZodArray<ZodSchema>>> {
  const shape: Record<string, ZodSchema | z.ZodArray<ZodSchema>> = {};

  for (const key in expandedFields) {
    const value = expandedFields[key];
    if (Array.isArray(value)) {
      // If it's an array, wrap the first element as a ZodArray (assuming all elements are the same schema)
      shape[key] = z.array(value[0]);
    } else {
      shape[key] = value;
    }
  }

  return z.object(shape);
}
