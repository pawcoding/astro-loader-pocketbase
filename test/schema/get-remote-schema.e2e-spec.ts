import { assert, beforeAll, describe, expect, it } from "vitest";
import { getRemoteSchema } from "../../src/schema/get-remote-schema";
import { getSuperuserToken } from "../../src/utils/get-superuser-token";
import { checkE2eConnection } from "../_mocks/check-e2e-connection";
import { createLoaderOptions } from "../_mocks/create-loader-options";

describe("getRemoteSchema", () => {
  const options = createLoaderOptions();
  let token: string;

  beforeAll(async () => {
    await checkE2eConnection();

    assert(options.superuserCredentials, "Superuser credentials are not set.");
    assert(
      !("impersonateToken" in options.superuserCredentials),
      "Impersonate token should not be used in tests."
    );

    const superuserToken = await getSuperuserToken(
      options.url,
      options.superuserCredentials
    );
    assert(superuserToken, "Superuser token should not be undefined");

    token = superuserToken;
  });

  it("should return undefined if superuser token is invalid", async () => {
    const result = await getRemoteSchema(options, "invalid-token");

    expect(result).toBeUndefined();
  });

  it("should return undefined if fetch request fails", async () => {
    const result = await getRemoteSchema(
      {
        ...options,
        collectionName: "invalidCollection"
      },
      token
    );

    expect(result).toBeUndefined();
  });

  it("should return schema if fetch request is successful", async () => {
    const result = await getRemoteSchema(
      {
        ...options,
        collectionName: "users"
      },
      token
    );

    assert(result, "Schema is not defined.");

    // Delete unstable properties (now properly typed with passthrough)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    delete (result as any).created;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    delete (result as any).updated;
    for (const field of result.fields) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      delete (field as any).id;
    }

    // Delete email templates (now properly typed with passthrough)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    delete (result as any).authAlert;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    delete (result as any).otp;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    delete (result as any).verificationTemplate;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    delete (result as any).resetPasswordTemplate;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    delete (result as any).confirmEmailChangeTemplate;

    expect(result).toMatchSnapshot();
  });
});
