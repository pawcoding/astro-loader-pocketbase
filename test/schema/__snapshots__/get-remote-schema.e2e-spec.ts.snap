// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getRemoteSchema > should return schema if fetch request is successful 1`] = `
{
  "fields": [
    {
      "hidden": false,
      "name": "id",
      "required": true,
      "type": "text",
    },
    {
      "hidden": true,
      "name": "password",
      "required": true,
      "type": "password",
    },
    {
      "hidden": true,
      "name": "tokenKey",
      "required": true,
      "type": "text",
    },
    {
      "hidden": false,
      "name": "email",
      "required": true,
      "type": "email",
    },
    {
      "hidden": false,
      "name": "emailVisibility",
      "required": false,
      "type": "bool",
    },
    {
      "hidden": false,
      "name": "verified",
      "required": false,
      "type": "bool",
    },
    {
      "hidden": false,
      "name": "name",
      "required": false,
      "type": "text",
    },
    {
      "hidden": false,
      "maxSelect": 1,
      "name": "avatar",
      "required": false,
      "type": "file",
    },
    {
      "hidden": false,
      "name": "created",
      "onCreate": true,
      "onUpdate": false,
      "type": "autodate",
    },
    {
      "hidden": false,
      "name": "updated",
      "onCreate": true,
      "onUpdate": true,
      "type": "autodate",
    },
  ],
  "id": "_pb_users_auth_",
  "name": "users",
  "type": "auth",
}
`;
